lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,423 @@
+""" Collection and Structure Module for pbench-audit-server
+"""
+
+import os
+import time
+
+from collections import OrderedDict
+
+
+permdict = {
+    "0": "---",
+    "1": "--x",
+    "2": "-w-",
+    "3": "-wx",
+    "4": "r--",
+    "5": "r-x",
+    "6": "rw-",
+    "7": "rwx",
+}
+
+
+def filepermissions(mode):
+    """Convert File permissions from numeric to symbolic.
+    from '755' --> 'rwxr-xr-x' """
+    fperm = ""
+    modebits = oct(mode)[-3:]
+    for bits in modebits:
+        if bits in permdict:
+            fperm += permdict[bits]
+        else:
+            raise ValueError(
+                f"Could not find key '{bits}' in file permissions dictionary"
+            )
+
+    return fperm
+
+
+class Hierarchy(object):
+    """Parent class of Hierarchies"""
+
+    def __init__(self, name, path, config):
+        self.name = name
+        self.path = path
+        self.config = config
+
+    def check_header(self, stat, fp, status):
+        """Check if we need to add header for Results or Incoming Hierarchy"""
+        if stat:
+            self.header(fp, status)
+        return
+
+    def header(self, fp, status):
+        """Adds main starting and Ending messages for each Hierarchy"""
+        lead_newline = "\n" if status == "start" else ""
+        fp.write(
+            f"{lead_newline}\n{status}-{self.config.TS[4:]}: {self.name} hierarchy: {self.path}\n"
+        )
+        return
+
+    def check_controller_in_list(self, controller, hierarchy_check):
+        """Checks whether there are controllers in dictionary or not"""
+        return any(
+            [controller in hierarchy_check[key]._dict for key in hierarchy_check]
+        )
+
+    def incoming_Results_dump(self, fp, stat, hierarchy_check):
```

Comment:
Why does the base class have a method that seems to indicate it knows about the details of the hierarchies?  Couldn't this just be called `dump()`?

Reply:
I will move it outside hierarchy.

Reply:
dump method is mainly used for common behaviour that all the four hierarchies show while this is only common between incoming and result.